### PR TITLE
[WIP] dialects/riscv: add linear scan register allocation strategy over basic block

### PR DESCRIPTION
**Pull request by adutilleul**
This PR introduces a basic version of the linear scan register allocation algorithm, which outperforms the current best register allocation strategy in the codebase.

The current approach limits live range computations to within the same basic block. However, future developments will enable cross-block computations (to handle loops and branching).

This improvement will lead to additional PRs by @compor and I, aimed at enhancing the algorithm's completeness and performance. For instance, incorporating a dependency like `sortedcontainers` could greatly benefit this algorithm and in general workset-related operations since the Python standard library lacks sorted collections.

### CODE REVIEW

File: tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// RUN: xdsl-opt -p riscv-allocate-registers{allocation_strategy=LinearScan} %s --print-op-generic | filecheck %s
+
+"builtin.module"() ({
+  %0 = "riscv.li"() {"immediate" = 6 : i32} : () -> !riscv.reg<>
```
Comment (tobiasgrosser) [inline tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir:4]: I feel this test case would very much benefit from custom syntax (https://github.com/xdslproject/xdsl/issues/1067).

Also, I am not fully sure what this test case is testing? It seems to contain a sequence of unused constants that are all assigned to the register. Does this cover any interesting corner case of the linar scan register allocator?

Would it make sense to add an option to run the register allocator with a limited register set, e.g., just 3, and use this to write smaller test cases?

File: xdsl/transforms/experimental/live_range.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                object.__setattr__(
+                    self, "start", owner.parent.get_operation_index(value.owner)
+                )
+
+                end = self.start
+                for use in value.uses:
+                    if parent_block := use.operation.parent:
+                        end = max(
+                            end,
+                            parent_block.get_operation_index(use.operation),
```
Comment (tobiasgrosser) [inline xdsl/transforms/experimental/live_range.py:45]: `get_operation_index` is a costly walk over the entire block.

File: xdsl/transforms/experimental/live_range.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            owner = value.owner
+
+            if isinstance(value, OpResult) and isinstance(owner.parent, Block):
+                object.__setattr__(self, "value", value)
+                object.__setattr__(
+                    self, "start", owner.parent.get_operation_index(value.owner)
+                )
+
+                end = self.start
+                for use in value.uses:
```
Comment (tobiasgrosser) [inline xdsl/transforms/experimental/live_range.py:41]: If we walk backward through the basic block, the first use that we find starts a live range. In that setting, I am actually curious if we even need a separate live-range data structure or if we could write the algorithm to just walk backwards. Walking backwards through the sequence of instructions makes the non-spilling variant of this algorithm easy:

for op in reverse(bb.ops):
  for operand in op.operands:
      reserve_register_if_not_already_reserved(operand)
  for res in op.results:
      free_register(res)

**Review by tobiasgrosser — CHANGES_REQUESTED** [review]
Some first comments. I have more coming later. Among others:

- Do you have test cases for live-range splitting?
- Would it make sense to have a spill/reload dialect?

File: tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir (hunk 1 above)
Comment (adutilleul) [reply tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir:4]: The test suite is yet to be improved indeed. The current file check was mainly here to ensure that there was not regression over the `block_naive` strategy (so not replacing already allocated registers).
The API for limiting the number of available registers is already there but not exposed yet as an attribute, but it will be used for future tests.

File: xdsl/transforms/experimental/live_range.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+@dataclass(init=False, frozen=True)
+class LiveRange:
+    """
+    Represents the live range of an SSA value as half-open interval using the
+    index of an Operation within its parent Block.
+    """
+
+    value: SSAValue
+    start: int
```
Comment (superlopuh) [inline xdsl/transforms/experimental/live_range.py:14]: Indices are really inefficient for dealing with linked lists, these should probably be ops instead

File: xdsl/transforms/experimental/live_range.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    end: int
+
+    def __init__(
+        self, value: SSAValue, start: int | None = None, end: int | None = None
+    ) -> None:
+        if start is not None and end is not None:
+            object.__setattr__(self, "value", value)
```
Comment (superlopuh) [inline xdsl/transforms/experimental/live_range.py:21]: ? why do you need this init

File: xdsl/transforms/riscv_register_allocation.py [hunk 6]
```diff
@@ -1,9 +1,63 @@
...
-from xdsl.ir import MLContext
+from xdsl.ir import MLContext, SSAValue
 from xdsl.passes import ModulePass
+from xdsl.transforms.experimental.live_range import LiveRange
+
+
+class RegisterSet:
+    def __init__(self, registers: list[str], reserved: set[str] = set()) -> None:
+        self.registers = registers
+        self.free: list[str] = list(registers)
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:15]: can you add these as field declarations on the class, please?

File: xdsl/transforms/riscv_register_allocation.py [hunk 7]
```diff
@@ -1,9 +1,63 @@
...
 from xdsl.dialects.builtin import ModuleOp
 from xdsl.dialects.riscv import Register, RegisterType, RISCVOp
-from xdsl.ir import MLContext
+from xdsl.ir import MLContext, SSAValue
 from xdsl.passes import ModulePass
+from xdsl.transforms.experimental.live_range import LiveRange
+
+
+class RegisterSet:
+    def __init__(self, registers: list[str], reserved: set[str] = set()) -> None:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:13]: init should probably take sequences

File: xdsl/transforms/experimental/live_range.py (hunk 5 above)
Comment (adutilleul) [reply xdsl/transforms/experimental/live_range.py:21]: If you want to pass manually computed live-ranges to test the linear scan algorithm without relying on the automatically computed live-ranges that's kinda necessary.

File: xdsl/transforms/experimental/live_range.py (hunk 4 above)
Comment (adutilleul) [reply xdsl/transforms/experimental/live_range.py:14]: At some point you need to compare the ranges for the LS algorithm, because right now you don't have any order relation on ops.

File: xdsl/transforms/riscv_register_allocation.py [hunk 8]
```diff
@@ -99,10 +286,11 @@ class RISCVRegisterAllocation(ModulePass):
 
     allocation_strategy: str = "GlobalJRegs"
 
-    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+    def apply(self, ctx: MLContext, op: ModuleOp, *args: Any) -> None:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:289]: I'm not sure this is a good idea, could ModulePass arguments be used instead?

Comment (adutilleul) [reply xdsl/transforms/riscv_register_allocation.py:289]: Some regalloc strategies might have different parameters, so with the current API, you can either all make them `ModulePass` or you dispatch unused `ModulePass` parameters to all strategies.

File: xdsl/transforms/riscv_register_allocation.py [hunk 9]
```diff
@@ -1,9 +1,63 @@
...
+        self.free = list(self.registers)
+        self.occupied = set()
+
+    def num_free_registers(self) -> int:
+        return len(self.free)
+
+    def has_available_registers(self) -> bool:
+        return not self.free
+
+    def limit_free_registers(self, limit: int) -> None:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:47]: can you document this, please?

File: xdsl/transforms/riscv_register_allocation.py [hunk 10]
```diff
@@ -89,6 +133,149 @@ def allocate_registers(self, module: ModuleOp) -> None:
                     self.idx += 1
 
 
+class RegisterLiveInterval(LiveRange):
+    """
+    Represents an enhanced live range with additional information for register allocation.
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:138]: Would be nice to expand this a lot, or add more inline comments. What's an abstract stack location? what's used?

File: xdsl/transforms/riscv_register_allocation.py [hunk 11]
```diff
@@ -89,6 +133,149 @@ def allocate_registers(self, module: ModuleOp) -> None:
...
+
+    def set_used(self, used: bool) -> None:
+        self.used = used
+
+
+class RegisterAllocatorLinearScan(AbstractRegisterAllocator):
+    """
+    Linear scan register allocation strategy.
+
+    Basic form of the algorithm based on the paper "Linear Scan Register Allocation"
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:175]: link?

File: xdsl/transforms/riscv_register_allocation.py [hunk 12]
```diff
@@ -89,6 +133,149 @@ def allocate_registers(self, module: ModuleOp) -> None:
...
+        self,
+        intervals: list[RegisterLiveInterval] | None = None,
+        register_set: RegisterSet = _DEFAULT_REGISTER_SET,
+    ) -> None:
+        self.active = OrderedDict()
+        self.register_set = register_set
+        self.intervals = intervals or []
+        self.idx_abstract_stack_location = 0
+
+    # TO:DO - Refactor the following methods to use a proper SortedSet (C++ fashion)
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:194]:
```suggestion
    # TODO: - Refactor the following methods to use a proper SortedSet (C++ fashion)
```
:)

File: xdsl/transforms/riscv_register_allocation.py (hunk 8 above)
Comment (superlopuh) [reply xdsl/transforms/riscv_register_allocation.py:334]: I don't get why they can't be stored on an instanec of this class

File: xdsl/transforms/riscv_register_allocation.py [hunk 13]
```diff
@@ -1,9 +1,80 @@
...
-from xdsl.ir import MLContext
+from xdsl.ir import MLContext, SSAValue
 from xdsl.passes import ModulePass
+from xdsl.transforms.experimental.live_range import LiveRange
+
+
+class RegisterSet:
+    # The complete set of registers
+    registers: Sequence[str]
+    # Registers that are free to be allocated
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:15]: nit: this would be useful as documentation comments so you see the doc when you hover over the usage

File: xdsl/transforms/experimental/live_range.py (hunk 4 above)
Comment (compor) [reply xdsl/transforms/experimental/live_range.py:14]: What does inefficient mean?
I doubt this will be a bottleneck unless we operate over hundreds of operations per block.
I'd rather avoid prematurely optimizing this.

Moreover, even LLVM's API uses numbering (`VNInfo` class), this is standard compiler treatment and it is way simpler to reason and handle numeric intervals, especially if splitting comes into play later.

File: xdsl/transforms/experimental/live_range.py (hunk 3 above)
Comment (compor) [reply xdsl/transforms/experimental/live_range.py:41]: I tried reversing `BlockOps` and it is not supported, so I ended up "caching" the operations in a list.
It felt clunkier unless we want to enhance that API.

Moreover, moving to spill/reload handling and exploring potential splitting (if required) soon.

File: xdsl/transforms/experimental/live_range.py (hunk 2 above)
Comment (compor) [reply xdsl/transforms/experimental/live_range.py:45]: `O(n)` indeed; I want something constant (or a caching version of this?) that allows for the creation of nice and tidy integer interval ranges that compilers have been using for decades in register allocation to compute very simple tasks like range splitting.

File: xdsl/transforms/experimental/live_range.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/transforms/experimental/live_range.py:14]: Why wouldn't we have hundreds of operations per block? It doesn't seem like a high number to me. We don't cache indices on operations, so this is O(n^2) in xDSL, which it isn't in LLVM/MLIR

File: xdsl/transforms/experimental/live_range.py (hunk 5 above)
Comment (superlopuh) [reply xdsl/transforms/experimental/live_range.py:21]: Sorry what I mean is the auto generated init would avoid this, I'd recommend doing further verification in `__post_init__(self)`

File: xdsl/transforms/riscv_register_allocation.py [hunk 14]
```diff
@@ -24,20 +78,11 @@ def allocate_registers(self, module: ModuleOp) -> None:
 
 class RegisterAllocatorBlockNaive(AbstractRegisterAllocator):
     idx: int
+    register_set: RegisterSet
 
-    def __init__(self) -> None:
+    def __init__(self, register_set: RegisterSet = _DEFAULT_REGISTER_SET) -> None:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:108]: Register set is mutable, right? I'd recommend allowing None here, and creating a new default register set if None

File: tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir (hunk 1 above)
Comment (tobiasgrosser) [reply tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir:4]: We also have support for pass arguments (added by @AntonLydike) that allow your filecheck test cases to set such local variables.

Comment (tobiasgrosser) [reply tests/filecheck/dialects/riscv/riscv_register_allocation_linear_scan.mlir:4]: @adutilleul, what do you think about custom syntax? I feel this could be a separate patch that will make this patch a lot nicer to review.

File: xdsl/transforms/experimental/live_range.py (hunk 2 above)
Comment (tobiasgrosser) [reply xdsl/transforms/experimental/live_range.py:45]: See the algorithm I proposed before.

File: xdsl/transforms/experimental/live_range.py (hunk 3 above)
Comment (tobiasgrosser) [reply xdsl/transforms/experimental/live_range.py:41]: > However, our immediate goal is to be able to allocate the PulpNN kernels and I feel this is the shortest path.

Are we in a rush? I feel taking some time to explore new ways to do register allocation that are enabled by MLIR could be nice.

Comment (superlopuh) [reply xdsl/transforms/experimental/live_range.py:41]: It would be nice to add `walk_reverse` everywhere we have `walk`, to do this well.

Comment (superlopuh) [reply xdsl/transforms/experimental/live_range.py:41]: One thing that partial register allocation would let us do is some constraint-based iterative system. Where we build a graph of all ssa values that are alive at the same time (or might be alive at the same time), and in a loop assign one of the unused registers to the most constricted unassigned registers. This approach wouldn't rely on indices at all, just a single walk forwards.

Comment (tobiasgrosser) [reply xdsl/transforms/experimental/live_range.py:41]: https://github.com/xdslproject/xdsl/pull/1115 adds walk_reverse().

Comment (tobiasgrosser) [reply xdsl/transforms/experimental/live_range.py:41]: https://github.com/xdslproject/xdsl/pull/1115 is now merged thanks to @superlopuh's quick review.

**Comment by AntonLydike** [comment]
This is a huge PR, would it be possible to split out smaller parts of this to make individual review easier?

**Comment by superlopuh** [comment]
What's the latest on this? Should we close this PR? I still think that it would be great to have a register allocator with live ranges one day, even if it's not useful for the microkernels we're working on right now. I've merged the `RegisterSet` already, just called it `RegisterQueue`, so the only thing left is the ranges thing. I think we should either close it or update it next week.
